### PR TITLE
Moves procs to a single global list/array

### PR DESCRIPTION
**Pull request by ike709**
First step to nuking `DreamPath`. Changes to `call()()` are getting punted to a follow-up PR.

Probably has broken/bad things in here but TestGame seems to work.

### CODE REVIEW

**Comment by Hinaichigo** [comment]
Would this fix certain procs not being recognized even though they are declared elsewhere?

**Comment by wixoaGit** [comment]
> Would this fix certain procs not being recognized even though they are declared elsewhere?

No, this just changes how existing procs are referenced internally. Could you open an issue for this?

File: DMCompiler/DM/DMProc.cs [hunk 1]
```diff
@@ -49,6 +49,8 @@ private class DMProcScope {
         public Location Location = Location.Unknown;
         public ProcAttributes Attributes;
         public string Name { get => _astDefinition?.Name; }
+        public int Id;
+        private DMObject _dmObject;
```
Comment (wixoaGit) [inline DMCompiler/DM/DMProc.cs:53]: Move this down with the rest of the private properties.

File: DMCompiler/DM/Visitors/DMObjectBuilder.cs [hunk 2]
```diff
@@ -13,14 +13,14 @@ class DMObjectBuilder {
             DMObjectTree.Reset();
             ProcessFile(astFile);
 
+
+            // TODO Nuke this pass
```
Comment (wixoaGit) [inline DMCompiler/DM/Visitors/DMObjectBuilder.cs:17]: Extra line here
```suggestion
            ProcessFile(astFile);

            // TODO Nuke this pass
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: DMCompiler/DM/DMObject.cs [hunk 3]
```diff
@@ -139,24 +128,19 @@ class DMObject {
...
-                    }
-
-                    typeJson.Procs.Add(procs.Key, procJson);
+            if (Procs.Count > 0)
+            {
+                typeJson.Procs = new List<List<int>>(Procs.Count);
+                foreach (var (_, value) in Procs)
+                {
+                    typeJson.Procs.Add(value);
                 }
```
Comment (wixoaGit) [inline DMCompiler/DM/DMObject.cs:142]:
```suggestion
                typeJson.Procs = new List<List<int>>(Procs.Values);
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**